client/src/components/overview/SizeSelector.jsx

```diff
@@ -1,50 +1,105 @@
 import React from 'react';
 
-const SizeSelector = (props) => {
-  const items = [];
-  for (let i = 0; i < props.skus.length; i++) {
-    items.push({
-      sku: props.skus[i],
-      size: props.sizes[i]
-    });
+class SizeSelector extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.onChange = this.onChange.bind(this);
+
+    this.state = {
+      currentSize: ''
+    }
+  }
+
+  componentDidUpdate() {
+    const items = this.getItems(this.props.skus, this.props.sizes);
+    const isInStock = (items && items.length > 0);
+    console.log('componentDidUpdate - Size Selector:');
+    console.log('isInStock', isInStock);
+    console.log('this.state.currentSize', this.state.currentSize);
+    console.log('this.props.sizes', this.props.sizes);
+
+    if (isInStock && this.state.currentSize !== '' && this.props.sizes.indexOf(this.state.currentSize) === -1) {
+      console.log('Resetting current size');
+      this.setState({ currentSize: '' });
+    }
+  }
+
+  getItems(skus, sizes) {
+    const items = [];
+    for (let i = 0; i < skus.length; i++) {
+      items.push({
+        sku: skus[i],
+        size: sizes[i]
+      });
+    }
+
+    return items;
+  }
+
+  resetToDefaultSelected() {
+    console.log('Resetting size selected')
+    const options = document.querySelectorAll('#po-sizes option');
+    for (let i = 0; i < options.length; i++) {
+      options[i].selected = options[i].defaultSelected;
+    }
   }
 
-  const onChange = (e) => {
-    props.onSelect(e?.target?.value);
+  onChange(e) {
+    const newSku = e?.target?.value;
+    const index = this.props.skus.indexOf(newSku);
+    if (index > -1) {
+      const newSize = this.props.sizes[index];
+      if (newSku) {
+        console.log('Size onChange:', newSize);
+        this.setState({ currentSize: newSize})
+        this.props.onSelect(newSku);
+      }
+    }
   };
 
-  let key = 0;
-  const isInStock = (items && items.length > 0);
-  const defaultMessage = isInStock ? 'Select Size' : 'Out of Stock';
-  return (
-    <div id="po-size-selector">
-      <select
-        id="po-sizes"
-        name="sizes"
-        defaultValue={ -1 }
-        onChange={ onChange }
-        disabled={ !isInStock }
-      >
-        <option
-          class="uppercase"
-          value={ -1 }
-          disabled
-          hidden
+  render() {
+    const items = this.getItems(this.props.skus, this.props.sizes);
```

Comment:
Same as prior comments: these values have already been determined. Consider using class variables to avoid re-calculating here.

Reply:
Oh, so the third state I found made this extra annoying, but the pattern I picked up for things like this that always have to be up to date based on props AND state:

1. Constructor, when first initializing
2. Render, when props are updated
3. ComponentDidUpdate, when state should be updated after initialization based on change in props - such as setting a starting value as a control changes from disabled to enabled based on new props. This forces a second rendering with the fully updated state of the component.

I hate these redundant calls, but alterable class variables also make me nervous. I need to find a solution!